// @buildcosmo/widget/dist/vite-plugin-cosmo.js
import { spawnSync } from 'child_process';
export function cosmo(options = {}) {
    return {
        name: 'vite-plugin-cosmo',
        apply: 'serve', // Only run in dev mode
        configureServer(server) {
            // Hook into server start
            server.httpServer?.once('listening', () => {
                if (!options.serverOnly) {
                    const widgetPath = options.widgetPath || server.config.root;
                    const devServerUrl = getDevServerUrl(server);
                    console.log('Creating a dev widget on the desktop. Ensure Cosmo is running and Developer Mode is enabled.');
                    notifyCosmo(widgetPath, devServerUrl);
                }
            });
        }
    };
}
function getDevServerUrl(server) {
    if (server.resolvedUrls?.local?.length) {
        return server.resolvedUrls.local[0];
    }
    const address = server.httpServer?.address();
    if (address && typeof address === 'object') {
        const protocol = server.config.server?.https ? 'https' : 'http';
        const host = address.address && address.address !== '::' ? address.address : 'localhost';
        return `${protocol}://${host}:${address.port}/`;
    }
    const fallbackPort = server.config.server?.port ?? 5173;
    const protocol = server.config.server?.https ? 'https' : 'http';
    return `${protocol}://localhost:${fallbackPort}/`;
}
function notifyCosmo(widgetPath, devServerUrl) {
    const cosmoUrl = buildCosmoURL(widgetPath, devServerUrl);
    const script = `
    tell application "Cosmo"
      activate
      open location "${cosmoUrl}"
    end tell
  `;
    const result = spawnSync('osascript', ['-e', script], { stdio: 'inherit' });
    if (result.status !== 0) {
        console.warn('Failed to run widget in dev mode. Make sure Cosmo is running and Developer Mode is enabled.');
    }
}
function buildCosmoURL(widgetPath, devServerUrl) {
    const params = new URLSearchParams({
        path: widgetPath,
        server: devServerUrl,
    });
    return `cosmo://devmode?${params.toString()}`;
}
